Issue 163 - Enable Request Response Logging

suryadipnag opened the pull request:
# Pull Request Review

## Checklist

- [x] Issue: https://github.com/IBM/ansible-lifecycle-driver/issues/163
- [x] List of related PRs : N.A.
- [x] Project builds without any errors.
- [x] Unit Tests - all pass.
- [x] Ran manual functional “smoke” test (does product as a whole still work?).
- [x] User Story meets the acceptance criteria - and passes. acceptance criteria should be understood at outset.
- [x]  **Description of the changes**:  

  We have to enable Request & Response logging with ExternalRequestId, ContentType, Payload and MessageDirection in Ansible Driver.
  

-    Created a new Ansible Module which will help us to capture the log data.
-    Implemented code changes in Ansible Driver which will print those captured logs.

suryadipnag committed:
Adding logging for ibm-cp4na-log-message ansible module

suryadipnag committed:
Adding ibm-cp4na-log-message module to the ansible module directory

suryadipnag committed:
Creating new ansible module ibm-cp4na-log-message

dvaccarosenna commented on docker/ibm-cp4na-log-message.py:
```diff
@@ -0,0 +1,47 @@
+from ansible.module_utils.basic import AnsibleModule
+import requests
+
+def run_module():
```
I think we probably need to unit test this Ansible module (use it in a playbook in a unit test) to verify it works and that we're getting the right result?

suryadipnag replied:
> I think we probably need to unit test this Ansible module (use it in a playbook in a unit test) to verify it works and that we're getting the right result?

Sure Dan @dvaccarosenna . We will add unit test for this module as you suggested. We have tested the new module with command line ansible-playbook and also tested that module with ansible driver from Nimrod UI with Assembly. We are getting the desired results for the module when we run the attached playbook script which is having our custom module. But adding the unit test for this module is definitely a good idea.

[test_script.txt](https://github.com/IBM/ansible-lifecycle-driver/files/9021921/test_script.txt)

dvaccarosenna reviewed (changes requested):
Is there a plan to add documentation for this module in the driver docs?

suryadipnag commented:
> Is there a plan to add documentation for this module in the driver docs?

Hi Dan @dvaccarosenna, Okay, we will add the documentation in the driver docs.

suryadipnag committed:
Update ibm-cp4na-log-message.py

suryadipnag committed:
Creating document for ibm-cp4na-log-message ansible module and it's usage in Ansible-Driver

Creating document for ibm-cp4na-log-message ansible module and it's usage in Ansible-Driver

suryadipnag committed:
Update ansible.py

Commit:
Adding unit test for ibm-cp4na-log-message module

suryadipnag committed:
Update test_ansible.py

Commit:
Changing the module location from docker to ansibledriver

Commit:
Merge remote-tracking branch 'origin/issue-163' into issue-163

Commit:
Changing the ansible custom module name and updating unit test

suryadipnag commented:
1. Added documentation for the new ansible module.
2. Renamed the module from ibm-cp4na-log-message to ibm_cp4na_log_message
3. Changed the directory of the module from 'docker' to 'ansibledriver' source code.
4. Added unit test for the new ansible module which will run a playbook script having ibm_cp4na_log_message module
5. Updated the Dockerfile

dvaccarosenna commented on ansibledriver/service/ansible.py:
```diff
@@ -331,6 +332,7 @@ def _log_unreachable_event(self, result):
             if delegated_vars is not None:
                 event.delegated_host_name = delegated_vars['ansible_host']
             self.event_logger.add(event)
+            self._generate_additional_logs(result)
```
In what scenario would an unreachable node require logging of requests? (the `ibm_cp4na_log_message` cannot have been called here). So not sure this call is needed here.

dvaccarosenna commented on ansibledriver/service/ansible.py:
```diff
@@ -406,6 +408,7 @@ def _log_event_for_failed_task(self, result, is_item=False):
             task_name = result._task.get_name().strip()
             event = TaskFailedOnHostEvent(task_name, host_name, result._result, item_label=item_label, delegated_host_name=delegated_host_name)
             self.event_logger.add(event)
+            self._generate_additional_logs(result)
```
In what scenario would a failed task require logging of requests? (the `ibm_cp4na_log_message` cannot have been called here or would have failed). So not sure this call is needed here.

haynesdavis replied:
should we keep it if customer wants to use `failed_when`, where task will fail but logs will be logged

dvaccarosenna replied:
It should be ok as the next task in their playbook should use `ibm_cp4na_log_message` which should only ever pass

dvaccarosenna commented on ansibledriver/service/ansible.py:
```diff
@@ -452,6 +455,7 @@ def _log_event_for_skipped_task(self, result, is_item=False):
             task_name = result._task.get_name().strip()
             event = TaskSkippedOnHostEvent(task_name, host_name, result._result, item_label=item_label, delegated_host_name=delegated_host_name)
             self.event_logger.add(event)
+            self._generate_additional_logs(result)
```
Same as above (is this call needed?)

dvaccarosenna commented on ansibledriver/service/ansible.py:
```diff
@@ -486,6 +490,7 @@ def _log_event_for_ok_task(self, result, is_item=False):
             task_name = result._task.get_name().strip()
             event = TaskCompletedOnHostEvent(task_name, host_name, result._result, item_label=item_label, delegated_host_name=delegated_host_name)
             self.event_logger.add(event)
+            self._generate_additional_logs(result)
```
Same as above, not sure this call is needed as the `ibm_cp4na_log_message` won't have been called here?

dvaccarosenna replied:
Doesn't this need to be called in `v2_runner_on_ok` for when the `ibm_cp4na_log_message` task completes ok?

haynesdavis replied:
`v2_runner_on_ok` and `v2_runner_item_on_ok` calls `_log_event_for_ok_task`. So, shouldn't adding `_generate_additional_logs` to `_log_event_for_ok_task` cover both?

dvaccarosenna replied:
Ok yeah I can see what you mean. I think that's fine then, it allows `ibm_cp4na_log_message` to be used in a loop which may be useful one day.

dvaccarosenna commented on tests/unit/service/test_ansible.py:
```diff
@@ -707,4 +709,61 @@ def test_run_lifecycle_with_outputs_of_different_types(self):
+            system_properties = PropValueMap({
+            })
+
+            dst = self.__copy_directory_tree(os.getcwd() + '/tests/resources/ansible_with_ibm_cp4na_log_message_module')
```
Should use `os.path.join` to create this path instead of `/` directly (to support all operating systems)

suryadipnag replied:
Sure, we will change this.

dvaccarosenna commented on tests/unit/service/test_ansible.py:
```diff
@@ -707,4 +709,61 @@ def test_run_lifecycle_with_outputs_of_different_types(self):
+            'request_id': request_id
+            })
+
+            self.assertLifecycleExecutionEqual(resp, LifecycleExecution(request_id, STATUS_COMPLETE, None, {}))
```
Is it possible to confirm the logs were created as expected? (Even if we need to use `unittest.mock` to mock the logger)

suryadipnag replied:
Please see the attached snapshot of the generated logs when I ran the unit test. I could see the expected logs.

![image](https://user-images.githubusercontent.com/106679486/178432120-a4ba4cf8-839c-4f7a-9df0-eca6b9030552.png)

dvaccarosenna replied:
Ok thanks. I was wondering if there was a way to verify in the unit test that the logs are coming through? Without having to manually check the output.

dvaccarosenna commented on ansibledriver/service/ansible.py:
```diff
@@ -539,6 +544,27 @@ def get_result(self):
+              message_direction = result._result['message_direction']
+              external_request_id = result._result['external_request_id']
+              content_type = result._result['content_type']
+              message_data = result._result['message_data']
```
I believe the design of this feature has some outstanding discussions going on, in particular the idea of allowing additional fields to be included here. I recommend you discuss this with Hari to confirm the outcome of those discussions, before we merge this functionality in.

suryadipnag replied:
Sure, I will discuss with Hari before we merge this.

Commit:
Addressing review comments

suryadipnag committed:
Updating ibm-cp4na-log-message.md

Commit:
Pushing code changes to support new design changes

Commit:
Merge remote-tracking branch 'origin/issue-163' into issue-163

suryadipnag commented:
Hi Dan @dvaccarosenna ,
Please find the below updates:
1) Code have been modified to accommodate the new changes made in our internal Design Document.
2) Updated the custom ansible module document in docs folder with the latest design.
3) Review comments have been addressed in the new code changes. Only one thing we have not changed yet but if you want we can change that too. We had added `_generate_additional_logs()` to `_log_event_for_ok_task()` intentionally so that it can cover both `v2_runner_on_ok()` & `v2_runner_item_on_ok()` as both call `_log_event_for_ok_task()`. Please refer to this link - https://github.com/IBM/ansible-lifecycle-driver/pull/164/files#r917869338
4) We have tested the new module with different scenarios. It is working fine. Below are some of the negative scenarios which we tested.
	a) If any one of the parameter (`message_direction`, `external_request_id `& `protocol`) is missing, task will fail as those 3 params are mandatory.
	b) If `message_direction` value is given other than `sent or received`, task will fail as it supports only 2 values, sent and received.
	c) If `message_data `is given but `content_type `is not given and vice versa, task will fail as both should exist together.
	etc ...
5) Tested in OCP cluster by creating Assembly Instances on Nimrod UI. We had pushed the resource package which contains playbook script having our custom module. We could see desired logs.
**OCP cluster log (Generated Request log)**
`{"@timestamp": "2022-07-14T13:09:30.299Z", "@version": "1", "message": "{'name': 'Surya', 'gender': 'male', 'email': 'surya2@test.com', 'status': 'active'}", "host": "ansible-lifecycle-driver-57bbbc5dd9-9wmtr", "path": "/usr/local/lib/python3.9/site-packages/ansibledriver/service/ansible.py", "tags": [], "type": "logstash", "thread_name": "MainThread", "level": "INFO", "logger_name": "ansibledriver.service.ansible", "tracectx.processid": "fbf83d27-f907-4d54-a679-84806b1c4844", "tracectx.transactionid": "3b38a365-a616-4936-8423-9db853cd8a3e", "tracectx.taskid": "13", "messageDirection": "sent", "tracectx.externalRequestId": "5d1cd9ca-f6d9-11ec-8084-00000a0b650b", "ContentType": "application/json", "protocol": "http", "protocol_metadata": {"url": "https://gorest.co.in/public/v2/users"}, "tracectx.driverrequestid": "64c7f66b-81f9-458b-8477-526513d03b15"}`
6) Unit test is also passing. Below is the Unit Test log snapshot.
<img width="764" alt="unit test log snapshot" src="https://user-images.githubusercontent.com/106679486/179041423-bb78a6ec-6030-4eb3-ae05-94043b3c2c33.png">

dvaccarosenna commented on ansibledriver/ibm_cp4na_log_message.py:
```diff
@@ -0,0 +1,58 @@
+    module_args = dict(
+        message_direction=dict(type='str', required=True, choices=['sent', 'received']),
+        external_request_id=dict(type='str', required=True),
+        message_data=dict(type='str'),
```
I know there have been some late design changes, would it be possible to get the new `message_type` field in here before we merge this? Let me know if it's too late to add that.

suryadipnag replied:
Sure. We have now added code for supporting `message_type` in the module. Code has been pushed.

dvaccarosenna commented on ansibledriver/ibm_cp4na_log_message.py:
```diff
@@ -0,0 +1,58 @@
+    if(protocol is None):
+        module.fail_json("Please provide a non empty protocol")
+    if(external_request_id is None):
+        module.fail_json("Please provide a non empty external_request_id")
```
`message_direction` and `message_type` (if we add it) should be mandatory as well

suryadipnag replied:
Yes, `message_direction` was already made mandatory in our Ansible module code. Now we have done the same for the newly added field `message_type` as well.

Commit:
Adding new field message_type

suryadipnag committed:
Updating the doc for adding new field message_type

suryadipnag commented:
Hi Dan @dvaccarosenna ,

We have done the code changes to support the additional field `message_type` in our Ansible module. After the code changes, we have executed unit test and it is passing. Also tested from Nimrod UI by creating Assembly instances and could see the expected logs including `message_type`.

dvaccarosenna commented on ansibledriver/service/ansible.py:
```diff
@@ -539,6 +541,39 @@ def get_result(self):
+              protocol = result._result['protocol']
+              protocol_metadata = result._result['protocol_metadata']
+
+              logging_context_dict = {'messageDirection' : message_direction, 'tracectx.externalRequestId' : external_request_id, 'ContentType' : content_type,
```
Sorry guys, just noticed the format of these fields are incorrect. It should be:
- `message_direction` instead of `messageDirection`
- `tracectx.externalrequestid` instead of `tracectx.externalRequestId`
- `content_type` instead of `ContentType`
- `message_type` instead of `messageType`

suryadipnag replied:
We have just now changed the string values as you have suggested @dvaccarosenna
Code changes have been pushed.

Commit:
Changing the key name in logs

Commit:
Merge remote-tracking branch 'origin/issue-163' into issue-163

suryadipnag committed:
Updating the key name of logstash logs
